### ISSUE TITLE
fix(models): dedicated root-cause debugging of the DeepSeek-V2-Lite greedy nondeterminism (follow-up to #817)

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#824]
> **UPDATE (root cause pinned):** A dedicated debugging pass has pinned this to the **MLX 0.32.1 bump `6c53673` (#704, 2026-07-09)** via a clean deterministic bisect. Key reframing of #817's conclusion: with `MLX_USE_CUDA_GRAPHS=0` the output is **deterministic garbage** (not a memory-safety nondeterminism), `compute-sanitizer` initcheck+memcheck are both clean, and the fault is an **upstream MLX CUDA kernel change** breaking `deepseek_v2`'s naive-MLA shape (qk 192 / v 128); `deepseek_v3`'s absorbed-MLA is unaffected. Full analysis and the concrete next steps (sub-bisect the upstream MLX range, then patch-overlay / upstream-report / absorbed-MLA-rewrite) are in the comment below. No fix merged yet; the investigation plan below still applies for the remaining upstream-narrowing work.

---

## Summary

This is the dedicated root-cause investigation for the DeepSeek-V2-Lite greedy nondeterminism established in #817. #817 determined that the "regressed since the benchmark" framing was wrong and that a mechanical bisect is not viable (there is no coherent-good anchor to bisect from), so a code fix was intentionally not forced. This issue owns the actual debugging: find and close the memory-safety defect that makes `deepseek_v2` produce different greedy output across identical fresh-process runs, then restore deterministic, coherent decoding on both backends.

## Established facts (from #817)

- **Nondeterministic greedy decode.** At `--temp 0`, repeated fresh-process runs of the same binary on the same prompt produce different outputs, most of them degenerate (a repeated single char like `!` or a repeated punctuation gram like `}'_` / `</:`). Greedy decode should be bit-identical run to run; it is not. That is a memory-safety class defect (uninitialized read, out-of-bounds, or a race), not a stable numeric error.
- **Backend-agnostic.** Reproduced on both CUDA and CPU (`MLXCEL_DEVICE=cpu`), so it is not a CUDA-kernel-only fault. The fused decode-MoE default-on (#282/#285) is an additional CUDA-specific degradation stacked on top (it drops coherence further), but it is not the root.
- **Pre-existing.** At the 2026-05-19 benchmark commit (331ac6e), coherence is already only ~50% (measured 8/12 and 5/12 coherent over fresh greedy runs with fused MoE off). The benchmark's ✅ means only "generates 100 tokens without crashing", not coherent output, which is why this went unnoticed.
- **Corruption is at the first generated token**, so it originates in prefill or weight loading, not late in decode.
- **Checkpoint is intact** (shard sha256 match HuggingFace; unmodified since January).
- **Specific to the `deepseek_v2` geometry.** Sibling models are stable and coherent on the same binary: qwen1.5-moe (switch_layers MoE, standard attention), youtu-vl (DeepSeek-V3-style MLA, with q_lora), unlimited-ocr (DeepSeek v1 MoE decoder). The distinguishing config of deepseek-v2-lite: `q_lora_rank = null` (no query compression, unlike V3 which always has q_lora), MLA head dims qk 192 = 128 nope + 64 rope / v 128, 64 routed + 2 shared experts, 6 active, `n_group = topk_group = 1`, `first_k_dense_replace = 1`.

## Leading hypothesis

The `q_lora_rank = null` branch of the MLA attention in `src/models/deepseek_v2.rs` is exercised by DeepSeek-V2-Lite but not by the working V3 siblings (which always compress the query), which makes it the prime suspect for an uninitialized/oversized scratch buffer, an incorrect mask/shape, or a KV-cache slice read before it is written. The MoE gather/expert-selection path (64 experts, 6 active, grouped routing degenerate at `n_group=1`) is the secondary suspect.

## Investigation plan

1. **Deterministic repro harness.** Reuse the coherence probe from the #817 investigation: N fresh greedy runs, classify each as coherent vs degenerate (repeated-char / repeated-gram / low-word-count), report a coherence rate. Establish a stable baseline rate on current `main` for both backends. This is the metric every subsequent step is measured against, because single runs are unreliable at ~50% flakiness.
2. **Confirm the corruption is in prefill.** Dump the first-token logits (or the top-k argmax) across several fresh runs. If they differ at `--temp 0`, prefill is nondeterministic and the search stays in prefill / weight-load; if they are identical but later tokens diverge, move the search into the decode step.
3. **Sanitizer sweep.** The CPU build makes this tractable: run the deepseek-v2-lite prefill under a sanitizer (`compute-sanitizer --tool initcheck/memcheck` for the CUDA path; ASan / Valgrind memcheck for the CPU build) to catch uninitialized reads and out-of-bounds directly. This is the highest-leverage step: a memory-safety defect that flips output run to run should surface here.
4. **Narrow to the `q_lora_rank = null` MLA branch.** Diff the attention forward in `src/models/deepseek_v2.rs` against the working V3 path (`deepseek_v3.rs`, used by youtu-vl). Look for a scratch/query buffer, a rope/nope split slice, an attention mask, or a KV-cache region that is allocated but read before it is fully written specifically when there is no q_lora projection.
5. **Rule the MoE path in or out.** Force a single-expert / dense configuration or a reference gather to see whether coherence goes to 100%, isolating the attention branch from the expert-gather branch.
6. **Verify the fix statistically.** A candidate fix must drive the probe to ~100% coherent across many reps on BOTH backends, and must make greedy output byte-identical across fresh processes. Do not accept a single coherent run as evidence (see #817: two runtime toggles each "fixed" it once by coincidence).

## Acceptance criteria

- Greedy (`--temp 0`) decode of `models/deepseek-v2-lite-4bit` is deterministic across fresh processes and coherent (correct answer to a simple prompt) on both CUDA and CPU, verified over many repetitions.
- A regression test that pins greedy determinism (identical output across two in-process generations with a fixed seed, or a coherence assertion) for a deepseek_v2 model, so this cannot silently regress again.
- The GB10 benchmark harness is updated to validate output coherence, not just "generates 100 tokens cleanly" (`docs/benchmark_results/model_tests_gb10.md` line 21). The false ✅ from a crash-free-but-garbage run is what hid this for two months; the harness should catch degenerate output.
- The corrected `docs/benchmark_results/model_tests_gb10.md` row for deepseek-v2-lite reflects reality once fixed.

## Refs

Root-cause follow-up to #817 (which holds the full measurement data and the corrected diagnosis). The fused decode-MoE default-on interaction is a separate, scoped degradation noted in #817; this issue targets the underlying nondeterminism.

**Comment by inureyes** [comment]
## Root cause pinned: the MLX 0.32.1 bump (6c53673 / #704)

A dedicated debugging pass has taken this from #817's "not bisectable, memory-safety mystery" to a pinned, deterministic root cause. Two findings reframe the bug, and a clean bisect names the exact commit.

### 1. It is deterministic, not a memory-safety nondeterminism

Running with `MLX_USE_CUDA_GRAPHS=0` (CUDA graph capture disabled) makes the output **fully deterministic**: 5/5 fresh greedy runs produce byte-identical garbage (`}'_}'_'_\...`). So the defect is really two layers stacked:

- A **deterministic numerical error** in the forward, present with graph capture off.
- **CUDA graph capture nondeterminism** layered on top (graph-on runs vary run-to-run), which is what made #817 read the whole thing as a memory-safety class bug and conclude it was not bisectable.

`compute-sanitizer` confirms it is not a memory-safety issue: with graphs off (so the sanitizer actually instruments the kernels), both `--tool initcheck` (uninitialized global reads) and `--tool memcheck` (out-of-bounds / misaligned) report **0 errors** while the garbage still reproduces. The forward is memory-safe but computes deterministically wrong values for the deepseek_v2 MLA shape.

### 2. The deterministic signal makes a clean bisect possible

- Anchor `331ac6e` (2026-05-19), graphs off: deterministic **coherent** (5/5 identical `8 times 6 equals 48.`).
- Current `main`, graphs off: deterministic **garbage** (5/5 identical).

`git bisect run` over `331ac6e..v0.4.1` with a deterministic graphs-off probe pinned the first bad commit:

**`6c53673` — chore: bump MLX pin to 0.32.1 (#703) (#704), 2026-07-09.**

Its only substantive change is the vendored MLX pin `e9463bbf` -> `57c66cac` (`src/lib/mlx-cpp/CMakeLists.txt`, `src/lib/mlxcel-core/build.rs`; the other hunk is a Metal-only overlay patch, irrelevant to CUDA). So the CUDA regression is entirely from upstream MLX kernel changes in the `e9463bbf..57c66cac` range.

### 3. Mechanism: an upstream CUDA kernel change breaks deepseek_v2's naive MLA shape

`deepseek_v2.rs`'s `MLAAttention` uses the **naive MLA** formulation: it materializes `keys = [k_nope, k_pe]` (head_dim 192 = 128 nope + 64 rope) and `values` (head_dim 128), then calls `attention_from_ptr` / `causal_attention` directly with asymmetric qk (192) vs v (128) dims. `deepseek_v3.rs` (used by youtu-vl, which is coherent on the same pin) uses the **absorbed/latent MLA** formulation instead (projects into latent space, passes the pe scores as an additive mask, projects the output back to v_head_dim), so it never feeds that asymmetric shape to the kernel. `deepseek_v2.rs`'s attention code is unchanged across the whole window (the only in-window edit is `c257af7`, additive VLM methods; the text forward is untouched), which is consistent with a shared-kernel regression rather than a model-code bug.

The upstream `57c66cac` bump summary names these candidate PRs in range: ml-explore/mlx#3804 (`Fix fp quantized matvec for output dim < 8`), ml-explore/mlx#3720 (`Fix compiled kernel correctness for negative-strided inputs`), ml-explore/mlx#3728 (Metal-only). The negative-strided-inputs one is a plausible suspect given the naive MLA path's heavy use of `slice_axis` (strided views) feeding the SDPA/rope kernels, but this is not yet confirmed.

### Next steps (this is where the follow-up work is)

1. Sub-bisect the vendored MLX range `e9463bbf..57c66cac` on the CUDA backend to name the exact upstream commit that breaks the qk=192 / v=128 naive-MLA path. Reuse the deterministic graphs-off probe against a fixed MLX checkout per step.
2. Once identified, choose the fix: (a) a targeted patch in mlxcel's vendored MLX overlay, (b) an upstream report to ml-explore/mlx plus a temporary mlxcel-side workaround, or (c) rewrite `deepseek_v2.rs` attention to the absorbed/latent MLA form that `deepseek_v3.rs` already uses (which is unaffected), sidestepping the asymmetric-shape kernel entirely. Option (c) is the most self-contained and testable in-tree, but needs careful numerical parity validation against the reference.
3. Separately, characterize the CUDA-graph-capture nondeterminism layer (why graph replay varies a deterministic forward) — it may be its own MLX issue.
4. Verify any fix statistically with the graphs-off deterministic probe (coherent on both CUDA and CPU across many reps) AND update the GB10 benchmark harness to check coherence, not just "generates 100 tokens" (`docs/benchmark_results/model_tests_gb10.md` line 21), so this class of regression is caught in CI rather than by chance.

### Reproduction commands used

```
# deterministic garbage (graphs off), current main:
MLX_USE_CUDA_GRAPHS=0 mlxcel generate -m models/deepseek-v2-lite-4bit -p "What is 7 times 8?" -n 20 --temp 0

# deterministic coherent at the anchor 331ac6e (graphs off).
# sanitizer clean under graphs off:
MLX_USE_CUDA_GRAPHS=0 compute-sanitizer --tool initcheck <same> ; --tool memcheck <same>  # 0 errors, garbage persists
```

No fix is merged: the root cause is now pinned to the upstream MLX 0.32.1 bump, but a verified fix requires the upstream sub-bisect and one of the options above, which is follow-up work. Keeping this open with the root cause and the concrete path. Refs #817, #703/#704 (the pinned bump).

**Comment by inureyes** [comment]
Fixed and merged in #829 (main `a9cc0c3`). Two layers:

1. **RMSNorm regression** (the root cause of the garbage): upstream MLX #3792 ("Fix CUDA RMSNorm small-row dispatch", first in the 0.32.1 pin) regressed the small-axis CUDA RMSNorm; DeepSeek-V2's `kv_a_layernorm` (axis 512, N_READS 8) hits the affected `block_dim==64` dispatch config. Fixed by keeping the MLX pin at the latest commit (`b7c3dd6d`) and overlaying `mlx/backend/cuda/rms_norm.cu` with the last-known-good pre-#3792 version.
2. **CUDA-graph-capture hazard** (pre-existing, the same class as Gemma 4 #688, and what dominated the server): extended the `MLX_USE_CUDA_GRAPHS=0` startup lever to the DeepSeek-V2 family.

End-to-end: `mlxcel-server -m deepseek-v2-lite-4bit` at default settings now answers coherently (5/5), where it returned garbage before; no regression on qwen3 / qwen2.5 / llama.

Upstream follow-up tracked in #830 (report the RMSNorm regression to ml-explore/mlx and drop the overlay once fixed).
